Vizr.php

```diff
@@ -6,4 +6,43 @@
 
 class Vizr extends AbstractExternalModule {
 
+  /**
+   * Override to check for Vizr User Rights. Else returns the parent value
+   * which checks for design rights.
+   */
+  public function redcap_module_link_check_display($project_id, $link) {
+    if ($this->hasVizrUserRights()) {
+      return $link;
+    } else {
+      return parent::redcap_module_link_check_display($project_id, $link);
+    }
+  }
+
+  /**
+   * Checks to see if a user has user rights for the Vizr external module on
+   * the current project.
+   */
+  public function hasVizrUserRights() {
+    $rights = \REDCap::getUserRights(USERID);
+    $user_module_config = $rights[USERID]['external_module_config'];
+    return !empty($rights) && is_array($user_module_config)
+        && in_array($this->PREFIX, $user_module_config);
+  }
+
+  /**
+   * Checks to see if a user has design rights for the current project.
+   */
+  public function hasDesignRights() {
+    $rights = \REDCap::getUserRights(USERID);
+    return !empty($rights) && $rights[USERID]['design'];
+  }
+
+  /**
+   * Checks to see if the user can edit charts. A user can edit charts if they
+   * are a super user or have design or user rights for the vizr module.
+   */
+  public function canEditVizrCharts() {
+    return SUPER_USER || $this->hasDesignRights() || $this->hasVizrUserRights();
```

Comment:
This potentially calls `getUserRights` twice so we might want to pass around `$rights` instead.

Reply:
I like the idea of fetching the rights once and just passing them to helpers that parse out what rights the user has.

Reply:
I don't fully understand the REDCap permissions model, but is this order of permission checks always correct? If you choose "Require module-specific user privilege", should design rights override that, or should these checks happen in the opposite order?

Reply:
That's a good question @heathharrelson . I think you are correct though. Module-rights should likely trump design rights, but that may be up for interpretation depending on how we design it. It might be a good idea to discuss this with @lizohsu too to see if she has any insight or expectation about the permission model. I could almost envisioin that we ditch the Design rights and just delegate back to Module rights.

Reply:
I tested this as-is on stage and it's working as currently designed. I'm going to ask a question on the REDCap community site to see how this should work. My feeling is that design rights should override user rights, but I'm not sure. That's a good question.